### PR TITLE
[PLGN-216] Cherwell

### PR DESCRIPTION
**Pull request by ablakley-r7**
### Description

Describe the proposed changes:

  - Update SDK to resolve SSL Errors

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``rapid7/insightconnect-python-3-38-slim-plugin:{sdk-version-num}`` and ``rapid7/insightconnect-python-3-38-plugin:{sdk-version-num}``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``insight-plugin validate`` which calls ``icon_validate`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `insight-plugin samples`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `insight-plugin validate` and make sure everything passes
2. Run the assessment tool: `insight-plugin run -A`. For single action validation: `insight-plugin run tests/{file}.json -A`
3. Copy (`insight-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

**Comment by ablakley-r7** [comment]
Plugin requires testing with a live Cherwell connection

File: plugins/cherwell/unit_test/util.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        if url == f"http://{STUB_URL}/CherwellAPI/api/V1/getbusinessobject/busobid/1/publicid/1":
+            return MockResponse(200, "lookup_incident_success.json.resp")
+        # Perform Ad Hoc Search
+        if url == f"http://{STUB_URL}/CherwellAPI/api/V1/getsearchresults":
+            return MockResponse(200, "perform_ad_hoc_search_success.json.resp")
+        # Update Incident
+        if url == f"http://{STUB_URL}/CherwellAPI/api/V1/SaveBusinessObject":
+            return MockResponse(200, "update_incident_success.json.resp")
+        # Error Handling
+        if url == f"http://{STUB_URL}/CherwellAPI/api/V1/getbusinessobject/busobid/1/publicid/400":
```
Comment (joneill-r7) [inline plugins/cherwell/unit_test/util.py:66]: could do a small tidy up here as well by iterating over the status codes and then subbing in that code to the URL string and mock response?

File: plugins/cherwell/Dockerfile [hunk 2]
```diff
@@ -1,25 +1,20 @@
-FROM komand/python-3-plugin:2
-# The three supported python parent images are:
-# - komand/python-2-plugin
-# - komand/python-3-plugin
-# - komand/python-pypy3-plugin
-#
-LABEL organization=komand
+FROM --platform=linux/amd64 rapid7/insightconnect-python-3-slim-plugin:5
```
Comment (joneill-r7) [inline plugins/cherwell/Dockerfile:1]: I think we're trying to adopt specifying the full version in this file to keep what we test/release on the same version?

File: plugins/cherwell/komand_cherwell/actions/create_incident/schema.py [hunk 3]
```diff
@@ -66,10 +67,7 @@ class CreateIncidentOutput(komand.Output):
       "order": 1
     }
   },
-  "required": [
-    "raw_response",
-    "success"
-  ]
+  "definitions": {}
```
[the remark this answers is not in the document]
Comment (ablakley-r7) [reply plugins/cherwell/komand_cherwell/actions/create_incident/schema.py:70]: I've set these all to non-required.

File: plugins/cherwell/komand_cherwell/util/api.py [hunk 4]
```diff
@@ -53,41 +56,38 @@ def _call_api(
...
+            self.logger.error(f"An error had occurred : {exception}" "If the issue persists please contact support")
+            raise PluginException(preset=PluginException.Preset.UNKNOWN, data=exception)
         except JSONDecodeError:
-            raise Exception(
-                f"Error: Received an unexpected response from {action_name}"
-                f"(non-JSON or no response was received). Response was: {resp.text}"
+            raise PluginException(preset=PluginException.Preset.INVALID_JSON, data=resp.text)
+
+    def response_handler(self, resp: Response, action_name: str = "", custom_error=None) -> None:
+        if custom_error and resp.status_code not in range(200, 299):
```
[the remark this answers is not in the document]
Comment (ablakley-r7) [reply plugins/cherwell/komand_cherwell/util/api.py]: Simplified the logic to only raise one exception but change the messages.

**Review by joneill-r7 — COMMENTED** [review]
Only one minor on a docstring.But code changes LGTM. 

Looking at the customer's error logs on the ticket is it that the certificate is invalid? Could be worth adding the ssl_verify into the connection? Although the host name being used when I try in my browser is valid but adding the API returns a strange XML error as is it's the wrong base hostname

File: plugins/cherwell/komand_cherwell/util/api.py [hunk 5]
```diff
@@ -52,49 +64,42 @@ def _call_api(
...
-            self.logger.error(f"An error had occurred : {e}" "If the issue persists please contact support")
-            raise
-
-        try:
+            resp = self.session.send(req, verify=self.ssl_verify)
+            self.response_handler(resp, action_name, custom_error)
             results = resp.json()
             return results
+        except requests.exceptions.HTTPError as exception:
+            self.logger.error(f"An error had occurred : {exception}" "If the issue persists please contact support")
```
Comment (joneill-r7) [inline plugins/cherwell/komand_cherwell/util/api.py:72]: typo, should be `has`?

File: plugins/cherwell/plugin.spec.yaml [hunk 6]
```diff
@@ -4,9 +4,15 @@ products: [insightconnect]
 name: cherwell
 title: Cherwell
 description: The Cherwell plugin is used to administrate incidents in Cherwell
-version: 2.1.1
+version: 3.0.0
+connection_version: 3
+sdk:
+  type: slim
+  version: 5.4.5
```
Comment (joneill-r7) [inline plugins/cherwell/plugin.spec.yaml:11]: sorry to be a pain, we should use 5.4.6 now seeing as this is the slim image.

File: plugins/cherwell/unit_test/util.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        action.connection = default_connection
+        action.logger = logging.getLogger("action logger")
+        return action
+
+    @staticmethod
+    def overwrite_params(dict_one, dict_two):
+        for key, value in dict_two.items():
+            if key in dict_one:
+                dict_one[key] = value
+        return dict_one
```
Comment (joneill-r7) [inline plugins/cherwell/unit_test/util.py:42]: nothing wrong with this (I think it essentially is just writing out what .update() is doing behind the scenes) but you could just do `dict_one.update(dict_two)` or or line 31 do `params.update(custom_params)` ?

**Review by joneill-r7 — COMMENTED** [review]
a couple of minor comments, but think we should update the SDK version that was released today! the rest LGTM

File: plugins/cherwell/komand_cherwell/connection/connection.py [hunk 8]
```diff
@@ -1,35 +1,43 @@
...
-        client_id = params["client_id"]["secretKey"]
-        authentication_mode = params["authentication_mode"]
+        client_id = params.get(Input.CLIENT_ID, {}).get("secretKey")
+        authentication_mode = params.get(Input.AUTHENTICATION_MODE)
+        ssl_verify = params.get(Input.SSL_VERIFY)
 
         # Form the base URL for the Cherwell server
-        scheme = "https://" if params["ssl_verify"] else "http://"
-        self._base_url = scheme + base_uri
+        if not (base_uri.startswith("http://") or base_uri.startswith("https://")):
```
[the remark this answers is not in the document]
Comment (ablakley-r7) [reply plugins/cherwell/komand_cherwell/connection/connection.py:26]: I don't think this should, if the uri doesn't start with https:// or http:// we'll move to prefix with https://

[the remark this answers is not in the document]
Comment (joneill-r7) [reply plugins/cherwell/komand_cherwell/connection/connection.py:26]: my bad, didn't see the `()` around the two conditions! 🤦

**Review by joneill-r7 — APPROVED** [review]
LGTM

**Pull request merged**